py/emitbc.c

```diff
@@ -213,34 +215,65 @@ STATIC void emit_write_bytecode_byte_child(emit_t *emit, int stack_adj, byte b,
     #endif
 }
 
-// unsigned labels are relative to ip following this instruction, stored as 16 bits
-STATIC void emit_write_bytecode_byte_unsigned_label(emit_t *emit, int stack_adj, byte b1, mp_uint_t label) {
+// Emit a jump opcode to a destination label.
+// The offset to the label is relative to the ip following this instruction.
+// The offset is encoded as either 1 or 2 bytes, depending on how big it is.
+// The encoding of this jump opcode can change size from one pass to the next,
+// but it must only ever decrease in size on successive passes.
+STATIC void emit_write_bytecode_byte_label(emit_t *emit, int stack_adj, byte b1, mp_uint_t label, bool is_signed) {
     mp_emit_bc_adjust_stack_size(emit, stack_adj);
-    mp_uint_t bytecode_offset;
-    if (emit->pass < MP_PASS_EMIT) {
-        bytecode_offset = 0;
-    } else {
-        bytecode_offset = emit->label_offsets[label] - emit->bytecode_offset - 3;
+
+    // Default to a 2-byte encoding (the largest) with an unknown jump offset.
+    unsigned int jump_encoding_size = 1;
+    ssize_t bytecode_offset = 0;
+
+    // Compute the jump size and offset only when code size is known.
+    if (emit->pass >= MP_PASS_CODE_SIZE) {
+        // The -2 accounts for this jump opcode taking 2 bytes (at least).
+        bytecode_offset = emit->label_offsets[label] - emit->bytecode_offset - 2;
+
+        // Check if the bytecode_offset is small enough to use a 1-byte encoding.
+        if ((is_signed && -64 <= bytecode_offset && bytecode_offset <= 63)
+            || (!is_signed && 0 <= bytecode_offset && bytecode_offset <= 127)) {
```

Comment:
Maybe this makes the intention a bit more obvious?

```suggestion
            || (!is_signed && (size_t)bytecode_offset < 256)) {
```

Reply:
It needs to fit in 7 bits to encode it as 1-byte.  So it could be:
```
|| (!is_signed && (size_t)bytecode_offset < 128)
```

Reply:
oops, yes I meant 128 instead of 256

Reply:
Updated.